better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1044 @@
+use access control, why we break programs into components like
+functions, types, and modules, and why we try to keep them small.
+
+## Components and Abstraction
+
+Obviously, “knowing something about `sort`” is only possible and
+relevant because `sort` is a separate component with its own name. If
+we had injected the sorting implementation directly into our code,
+we'd have to reason about all of its details, and we'd need a comment
+to tell us that the implementation was sorting the elements.
+
+So it's a common fallacy that factoring out a component only makes
+sense if it's used more than once. The real criterion is whether you
+can identify a unit of *abstraction*—the third ingredient of Better
+Code.  If you can think of a simple and descriptive name for
+something, that's usually a good indicator that there's an abstraction
+waiting to be discovered.
+
+In fact, creating a component directly creates the opportunity for an
+additional use site: in testing.  The more code you test separately,
+the better it is for correctness.  The more you can use understandable
+abstractions, the simpler and clearer the use-sites become, which is
+better for correctness. Abstraction and correctness support one
+another in a virtuous cycle.  And abstraction is necessary for
+programming at scale: because we can't keep the whole program in our
+heads, we do what humans always do when faced with complexity: we
+break complicated problems into parts that can be understood in
+isolation.
+
+## Hoare Logic
+
+Programming by contract grew out of Hoare Logic, developed in 1969 by
+the British computer scientist and logician Tony Hoare.  While we
+won't be using Hoare Logic directly, we'll discuss it briefly to
+provide historical context, and because it is the source of crucial
+terminology.
+
+### Preconditions and Postconditions
+
+Hoare used this notation, called a “Hoare triple,”
+
+> {P}S{Q}
+
+which is an assertion that if **precondition** *P* is met, operation
+*S* establishes **postcondition** *Q*.
+
+For example:
+
+- if we start with `x == 2` (precondition), after `x += 1`, `x == 3` (postcondition):
+
+  > {x == 2}x+=1{x == 3}
+
+- if `x` is less than the maximum integer (precondition), after `x
+  += 1`, `x` is greater than the minimum integer (postcondition):
+
+  > {x < Int.max}x+=1{x > Int.min}
+
+What makes preconditions and postconditions useful for formal proofs
+is this composition rule:
+
+> {P}S{Q} ∧ {Q}T{R} ⇒ {P}S;T{R}
+
+Given two valid Hoare triples, if the postconditions of the first are
+preconditions of the second, we can form a new valid triple describing
+the effects of performing the operations in sequence.  The notation
+may look exotic, but its meaning should be familiar: this rule simply
+captures the reasoning we use to think informally about statements
+when we write one statement after the next.
+
+### Invariants
+
+Not all code runs in a straight line, though, so Hoare also gave us a
+tool for reasoning about loops.
```

Comment:
We should state the While rule formally. You may want to mention the While rule for total correctness as well. The formalization is long, introducing the idea of a loop variant _t_ that is strictly decreasing with a finite domain. to prove termination. Then we would just be missing the Conditional rule to cover Hoare completely. If this is getting to be too much, we could mention that we will cover these in more detail in the Algorithm chapter (which might be a better place).

Reply:
I am dubious that we want to get into the While rule for total correctness at all.  The book is supposed to be small-ish and practical. The details of Hoare logic are much less powerful for everyday programming than DbC is and it's mostly introduced here to show that the formal reasoning one does with Hoare is on the continuum with normal activities.  The conditional rule is pretty obvious and I don't think a complete coverage of Hoare is going to benefit our readers, and might well turn lots of people off. I do think loop invariants are especially important for algorithms, and I figured you'd cover that in the algorithms chapter, since loops are so central.

Reply:
@sean-parent I'd like to resolve this; do we need to debate this point?

Reply:
> I don't think a complete coverage of Hoare is going to benefit our readers, and might well turn lots of people off

I agree with this. I presume there are more specialized works available that can be referenced for folks who are interested in a deeper treatment. I'm concerned that every time we introduce a formalism, we lose readers that would otherwise benefit from the rest of this chapter.